File: java/src/main/java/codingdojo/HeroQuest.java
package codingdojo;

import java.util.Objects;

public class HeroQuest {

    public static String playerToString(String playerName, int playerHealth, int playerStrength, int playerMagic, int playerCraftingSkill) {
        return String.format("%s's Attributes:\nHealth: %d\nStrength: %d\nMagic: %d\nCrafting Skill: %d\n",
                playerName, playerHealth, playerStrength, playerMagic, playerCraftingSkill);
    }

    public static void playerFallsDown(QuestData questData) {
        System.out.println("Player drops off a cliff.");

        if (questData.getPlayerStrength() < 5) {
            questData.setPlayerHealth(questData.getPlayerHealth() - 10);
            System.out.println("Player's strength is too small. Health decreases by 10.");
        }
    }

    public static String itemToString(String itemName, String itemKind, int itemPower) {
        return String.format("Item: %s\nKind: %s\nPower: %d\n", itemName, itemKind, itemPower);
    }

    public static void itemReduceByUsage(QuestData questData) {
        System.out.println(String.format("Using the item with kind '%s' and power %d", questData.getItemKind(), questData.getItemPower()));

        questData.setItemPower(questData.getItemPower() / 2);

        if (questData.getItemPower() == 0) {
            questData.setItemKind("Junk");
        }

    }

    public static void itemApplyEffectToPlayer(QuestData data) {
        System.out.println(String.format("Applying the effect of %s (%s):", data.getItemName(), data.getItemKind()));

        if (Objects.equals(data.getItemKind(), "Health")) {
            data.setPlayerHealth(data.getPlayerHealth() + data.getItemPower());
        } else if (Objects.equals(data.getItemKind(), "Strength")) {
            data.setPlayerStrength(data.getPlayerStrength() + data.getItemPower());
        } else if (Objects.equals(data.getItemKind(), "Magic")) {
            data.setPlayerMagic(data.getPlayerMagic() + data.getItemPower());
        } else {
            // ignore unknown item kind
        }
    }

    public static void itemRepair(QuestData questData) {
        System.out.println("Using the repair skill to fix the item:");

        int repairAmount = -5 + ((questData.getPlayerCraftingSkill() * 2) + 1);

        questData.setItemPower(questData.getItemPower() + repairAmount);

        System.out.println(String.format("Repaired the item by %d points. Item's Durability: %d", repairAmount, questData.getItemPower()));
    }
}

File: csharp/HeroQuest/HeroQuest.cs
namespace HeroQuest;

public class HeroQuest
{
    public static string PlayerToString(String playerName, int playerHealth, int playerStrength, int playerMagic, int playerCraftingSkill)
    {
        return string.Format("{0}'s Attributes:\nHealth: {1}\nStrength: {2}\nMagic: {3}\nCrafting Skill: {4}\n",
                playerName, playerHealth, playerStrength, playerMagic, playerCraftingSkill);
    }

    public static void PlayerFallsDown(QuestData questData)
    {
        Console.WriteLine("Player drops off a cliff.");

        if (questData.PlayerStrength < 5)
        {
            questData.PlayerHealth = questData.PlayerHealth - 10;
            Console.WriteLine("Player's strength is too small. Health decreases by 10.");
        }
    }

    public static string ItemToString(String itemName, String itemKind, int itemPower)
    {
        return string.Format("Item: {0}\nKind: {1}\nPower: {2}\n", itemName, itemKind, itemPower);
    }

    public static void ItemReduceByUsage(QuestData questData)
    {
        Console.WriteLine(string.Format("Using the item with kind '%s' and power %d", questData.ItemKind, questData.ItemPower));

        questData.ItemPower = questData.ItemPower / 2;

        if (questData.ItemPower == 0)
        {
            questData.ItemKind = "Junk";
        }

    }

    public static void ItemApplyEffectToPlayer(QuestData data)
    {
        Console.WriteLine(string.Format("Applying the effect of %s (%s):", data.ItemName, data.ItemKind));

        if (data.ItemKind == "Health")
        {
            data.PlayerHealth = data.PlayerHealth + data.ItemPower;
        }
        else if (data.ItemKind == "Strength")
        {
            data.PlayerStrength = data.PlayerStrength + data.ItemPower;
        }
        else if (data.ItemKind == "Magic")
        {
            data.PlayerMagic = data.PlayerMagic + data.ItemPower;
        }
        else
        {
            // ignore unknown item kind
        }
    }

    public static void ItemRepair(QuestData questData)
    {
        Console.WriteLine("Using the repair skill to fix the item:");

        int repairAmount = -5 + ((questData.PlayerCraftingSkill * 2) + 1);

        questData.ItemPower = questData.ItemPower + repairAmount;

        Console.WriteLine(string.Format("Repaired the item by %d points. Item's Durability: %d", repairAmount, questData.ItemPower));
    }
}
